### PR TITLE
Add userspace rverst

### PR DESCRIPTION
**Pull request by rverst**
## Description
This PR adds my userspace and a layout for the keyboards I use (id80)

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/id80/keymaps/rverst/keymap.c [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#include "rverst.h"
```
Comment (drashna) [inline keyboards/id80/keymaps/rverst/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: users/rverst/config.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#pragma once
```
Comment (drashna) [inline users/rverst/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: users/rverst/config.h [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+#define UNICODE
```
Comment (drashna) [inline users/rverst/config.h:3]: Not sure this is needed. 
```suggestion
```

File: users/rverst/config.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+#    define RGBLIGHT_SLEEP
+#    ifdef RGBLIGHT_ANIMATIONS
+#        undef RGBLIGHT_ANIMATIONS
+#    endif
+#    define RGBLIGHT_SLEEP
+#    define RGBLIGHT_EFFECT_BREATHING
+#    define RGBLIGHT_EFFECT_RAINBOW_MOOD
+#    define RGBLIGHT_EFFECT_RAINBOW_SWIRL
+#    define RGBLIGHT_EFFECT_STATIC_GRADIENT
+#    define TWINKLE
```
Comment (drashna) [inline users/rverst/config.h:15]:
```suggestion
#    define RGBLIGHT_EFFECT_TWINKLE
```

File: users/rverst/rverst.c [hunk 5]
```diff
@@ -0,0 +1,392 @@
+#include "rverst.h"
```
Comment (drashna) [inline users/rverst/rverst.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: users/rverst/rverst.c [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+#include "print.h"
+
+#ifdef UNICODE
```
Comment (drashna) [inline users/rverst/rverst.c:4]:
```suggestion
#ifdef UNICODEMAP_ENABLE
```

File: users/rverst/rverst.c [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+#    include "unicode.h"
+#endif
+
+userspace_config_t userspace_config;
+
+int get_mode(void) {
```
Comment (drashna) [inline users/rverst/rverst.c:10]:
```suggestion
uint8_t get_mode(void) {
```

File: users/rverst/rverst.c [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+        m += 2;
+    }
+    if (userspace_config.mode_3) {
+        m += 4;
+    }
+
+    return m;
+}
+
+void set_mode(int mode, bool save) {
```
Comment (drashna) [inline users/rverst/rverst.c:25]:
```suggestion
void set_mode(uint8_t mode, bool save) {
```

File: users/rverst/rverst.c [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+    } else {
+        userspace_config.mode_1 = false;
+    }
+
+    if (save) {
+        eeconfig_update_user(userspace_config.raw);
+    }
+}
+
+void switch_mode(int mode) {
```
Comment (drashna) [inline users/rverst/rverst.c:60]:
```suggestion
void switch_mode(uint8_t mode) {
```

File: users/rverst/rverst.c [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+#ifdef UNICODE
```
Comment (drashna) [inline users/rverst/rverst.c:61]:
```suggestion
#ifdef UNICODEMAP_ENABLE
```

File: users/rverst/rverst.c [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+            set_unicode_input_mode(UC_WINC);
+            break;
+        case LINUX_UNI:
+            set_unicode_input_mode(UC_LNX);
+            break;
+    }
+#endif
+}
+
+bool is_unicode(int mode) { return (mode == MAC_UNI) || (mode == WINDOWS_UNI) || (mode == LINUX_UNI); }
```
Comment (drashna) [inline users/rverst/rverst.c:76]:
```suggestion
bool is_unicode(uint8_t mode) { return (mode == MAC_UNI) || (mode == WINDOWS_UNI) || (mode == LINUX_UNI); }
```

File: users/rverst/rverst.c [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+
+//**********************
+// process_record
+//**********************
+__attribute__((weak)) bool process_record_keymap(uint16_t keycode, keyrecord_t *record) { return true; }
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    if (keycode < SAFE_RANGE) {
+        return true;
+    }
```
Comment (drashna) [inline users/rverst/rverst.c:127]: I wouldn't recommend this. You can alter the behavior of existing keycodes, if you want/need. 
```suggestion
```

File: users/rverst/rverst.c [hunk 13]
```diff
@@ -0,0 +1,392 @@
...
+
+        // Superscript three - ³
+        case RV_SUP3:
+            if (is_unicode(mode)) {
+                send_unicode_string("³");
+            }
+            return false;
+    }
+
+    return process_record_keymap(keycode, record);
```
Comment (drashna) [inline users/rverst/rverst.c:391]:
```suggestion
    return true;
```

File: users/rverst/rverst.c [hunk 14]
```diff
@@ -0,0 +1,392 @@
...
+__attribute__((weak)) bool process_record_keymap(uint16_t keycode, keyrecord_t *record) { return true; }
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    if (keycode < SAFE_RANGE) {
+        return true;
+    }
+
+    if (!record->event.pressed) {
+        return process_record_keymap(keycode, record);
+    }
```
Comment (drashna) [inline users/rverst/rverst.c:131]:
```suggestion
    if (!process_record_keymap(keycode, record)) { return false; }
```

File: users/rverst/rverst.c [hunk 15]
```diff
@@ -0,0 +1,392 @@
...
+
+    bool ls = keyboard_report->mods & MOD_BIT(KC_LSFT);
+    bool rs = keyboard_report->mods & MOD_BIT(KC_RSFT);
```
Comment (drashna) [inline users/rverst/rverst.c:134]:
```suggestion
    bool ls = get_mods() & MOD_BIT(KC_LSFT);
    bool rs = get_mods() & MOD_BIT(KC_RSFT);
```
or, even: 
```suggestion
    bool ls = (get_mods()|get_weak_mods()) & MOD_BIT(KC_LSFT);
    bool rs = (get_mods()|get_weak_mods()) & MOD_BIT(KC_RSFT);
```

File: users/rverst/rverst.h [hunk 16]
```diff
@@ -0,0 +1,66 @@
+#ifndef RVERST_H
```
Comment (drashna) [inline users/rverst/rverst.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: users/rverst/rverst.h [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+#define RVERST_H
```
Comment (drashna) [inline users/rverst/rverst.h:2]:
```suggestion
#pragma once
```

File: users/rverst/rverst.h [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+void set_mode(int mode, bool save);
+void switch_mode(int mode);
+bool is_unicode(int mode);
+
+void keyboard_pre_init_keymap(void);
+void keyboard_post_init_keymap(void);
+void eeconfig_init_keymap(void);
+bool process_record_keymap(uint16_t keycode, keyrecord_t *record);
+
+#endif  // RVERST
```
Comment (drashna) [inline users/rverst/rverst.h:66]:
```suggestion
```

File: users/rverst/rverst.h [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+        bool mode_3 : 1;
+    };
+} userspace_config_t;
+
+extern userspace_config_t userspace_config;
+
+int  get_mode(void);
+void set_mode(int mode, bool save);
+void switch_mode(int mode);
+bool is_unicode(int mode);
```
Comment (drashna) [inline users/rverst/rverst.h:59]:
```suggestion
uint8_t get_mode(void);
void    set_mode(uint8_t mode, bool save);
void    switch_mode(uint8_t mode);
bool    is_unicode(uint8_t mode);
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/id80/keymaps/rverst/keymap.c (hunk 1 above)
Comment (rverst) [reply keyboards/id80/keymaps/rverst/keymap.c:1]: Of course.

File: users/rverst/config.h (hunk 2 above)
Comment (rverst) [reply users/rverst/config.h:1]: Of course.

File: users/rverst/unicode.h [hunk 20]
```diff
@@ -0,0 +1,17 @@
+#ifndef RV_UNICODE_H
+#define RV_UNICODE_H
```
Comment (drashna) [inline users/rverst/unicode.h:2]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

And:
```suggestion
#pragma once
```

File: users/rverst/unicode.h [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+
+enum unicode_names { BANG, IRONY, SNEK };
+
+const uint32_t PROGMEM unicode_map[] = {
+    [BANG]  = 0x203D,
+    [IRONY] = 0x2E2E,
+    [SNEK]  = 0x1F40D,
+};
+
+#endif  // RV_UNICODE_H
```
Comment (drashna) [inline users/rverst/unicode.h:17]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: users/rverst/unicode.h (hunk 20 above)
Comment (rverst) [reply users/rverst/unicode.h:2]: Sure, sorry, missed that file.

**Review by drashna — APPROVED** [review]

File: users/rverst/license [hunk 22]
```diff
@@ -0,0 +1,14 @@
+Copyright 2021 Robert Verst <robert@verst.eu> @rverst
```
Comment (fauxpark) [inline users/rverst/license:1]: This file can be deleted.

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Pull request merged**